File: Solutions/85-maximal-rectangle/maximal-rectangle.rb
# We are given a binary matrix of characters '0' and '1'
# Our goal is to find the area of the largest rectangle containing only '1's

def maximal_rectangle(matrix)
  # Edge case: empty matrix
  return 0 if matrix.nil? || matrix.empty? || matrix[0].empty?

  rows = matrix.length
  cols = matrix[0].length

  # This array will store the "height" of consecutive 1's for each column
  # Think of it as a histogram that we build row by row
  heights = Array.new(cols, 0)

  # This will store the maximum rectangle area found so far
  max_area = 0

  # Iterate over each row of the matrix
  (0...rows).each do |r|
    # Update the histogram heights for this row
    (0...cols).each do |c|
      if matrix[r][c] == '1'
        # If current cell is '1', increase height
        heights[c] += 1
      else
        # If current cell is '0', reset height to 0
        heights[c] = 0
      end
    end

    # After updating heights, compute the largest rectangle
    # in the histogram represented by heights[]
    max_area = [max_area, largest_rectangle_area(heights)].max
  end

  max_area
end

# Helper function to compute the largest rectangle in a histogram
# This uses a monotonic increasing stack
def largest_rectangle_area(heights)
  # Stack will store indices of the histogram bars
  stack = []

  max_area = 0

  # We append a 0 at the end to force processing of all bars
  extended_heights = heights + [0]

  extended_heights.each_with_index do |h, i|
    # While the current height is less than the height
    # at the top of the stack, we can form rectangles
    while !stack.empty? && h < extended_heights[stack[-1]]
      # Height of the rectangle is the height of the popped bar
      height = extended_heights[stack.pop]

      # Width calculation:
      # If stack is empty, rectangle spans from index 0 to i-1
      # Otherwise, spans from stack[-1] + 1 to i - 1
      width = stack.empty? ? i : i - stack[-1] - 1

      # Update maximum area
      max_area = [max_area, height * width].max
    end

    # Push current index onto the stack
    stack << i
  end

  max_area
end

File: Solutions/85-maximal-rectangle/maximal-rectangle.kt
class Solution {
    fun maximalRectangle(matrix: Array<CharArray>): Int {

        // If the matrix is empty, there is no rectangle.
        if (matrix.isEmpty() || matrix[0].isEmpty()) {
            return 0
        }

        val rows = matrix.size
        val cols = matrix[0].size

        // heights[col] represents the number of consecutive
        // '1's ending at the current row in this column.
        //
        // Example:
        //
        // matrix:
        // 1 0 1
        // 1 1 1
        //
        // After processing the second row:
        //
        // heights:
        // 2 1 2
        //
        // This can now be treated as a histogram.
        val heights = IntArray(cols)

        // Keep track of the largest rectangle found so far.
        var maxArea = 0

        // Process the matrix one row at a time.
        for (row in 0 until rows) {

            // Update the histogram heights for this row.
            for (col in 0 until cols) {

                if (matrix[row][col] == '1') {

                    // This cell continues the vertical sequence
                    // of 1s from the previous row.
                    heights[col]++

                } else {

                    // A zero breaks the sequence.
                    // Therefore, there can be no rectangle
                    // containing this column at this height.
                    heights[col] = 0
                }
            }

            // Once we have the histogram for this row,
            // find the largest rectangle inside it.
            maxArea = maxOf(
                maxArea,
                largestRectangleInHistogram(heights)
            )
        }

        return maxArea
    }

    private fun largestRectangleInHistogram(heights: IntArray): Int {

        // The stack stores indices of bars.
        //
        // The heights of the bars represented by the stack
        // are maintained in increasing order.
        //
        // This allows us to quickly determine the width of
        // a rectangle when we encounter a smaller bar.
        val stack = java.util.ArrayDeque<Int>()

        // We will use an extra iteration with a height of 0.
        //
        // This forces all remaining bars in the stack to be
        // processed at the end.
        var maxArea = 0

        for (i in 0..heights.size) {

            // When i == heights.size, we are outside the
            // actual array.
            //
            // Treat this imaginary bar as height 0 so that
            // all remaining bars are removed from the stack.
            val currentHeight =
                if (i == heights.size) {
                    0
                } else {
                    heights[i]
                }

            // If the current bar is smaller than the bar
            // at the top of the stack, the taller bar can
            // no longer extend to the right.
            //
            // Therefore, we calculate its largest possible
            // rectangle now.
            while (stack.isNotEmpty() &&
                heights[stack.peek()] > currentHeight
            ) {

                // Index of the bar whose rectangle we are
                // calculating.
                val heightIndex = stack.pop()

                // Height of the rectangle.
                val height = heights[heightIndex]

                // After removing heightIndex from the stack:
                //
                // - The new stack top is the first smaller
                //   bar on the LEFT.
                //
                // - i is the first smaller bar on the RIGHT.
                //
                // Therefore, the rectangle extends between
                // those two boundaries.
                val width =
                    if (stack.isEmpty()) {
                        i
                    } else {
                        i - stack.peek() - 1
                    }

                // Calculate the area of the rectangle.
                val area = height * width

                // Keep the largest area found.
                maxArea = maxOf(maxArea, area)
            }

            // Add the current index to the stack.
            //
            // The stack remains monotonic because all taller
            // bars have already been removed above.
            if (i < heights.size) {
                stack.push(i)
            }
        }

        return maxArea
    }
}